test/test_nn.py

```diff
@@ -16431,6 +16431,123 @@ def _helper(zero_infinity):
         _helper(zero_infinity=True)
         _helper(zero_infinity=False)
 
+    def _CTCLoss_no_batch_dim_helper(self, device, input_length, vocab_size, target_length):
```

Comment:
The test logic is a bit tricky to validate since there's a lot going on. Rather than iterate through the various cases within a single test, I think it could be simplified quite a bit by doing a couple things:
* Write test logic for a single case (create unbatched & batched inputs, run both through, compare) and using `@parametrize_test` to iterate over the cases instead. I can definitely help with this if needed!
* No need to explicitly duplicate cases for the cudnn case. We usually write a CUDA test and assume that CI will run it in both cuDNN and non-cuDNN environments. If there is cudnn-specific stuff that needs to be done (e.g. place targets on CPU), you can check if cuDNN is enabled and do it then.
* You can also use `@parametrize_test("reduction", ["mean", "sum", "none"])` so you don't need to iterate over reductions in the test.

Reply:
- Incorporated `@parametrize_test` 
- Consolidated the CPU and CUDA tests by refactoring a bit and then split the shape checking / value checking tests
I think these together make it indeed look much cleaner!